#15961 issue:Allow Contact to unsubscribe from specific lists

Anurag810 opened the pull request:
Please read the [Pull Request Checklist](https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist) to ensure you have everything that is needed to get your contribution merged.

Commit:
intial commit

Commit:
changes

Commit:
added validation

rmehta commented on frappe/www/unsubscribed.py:
```diff
@@ -0,0 +1,20 @@
+		if "user_email" in frappe.form_dict:
+			email = frappe.form_dict['user_email']
+			email_group = get_email_groups(email)
+			print("---------------------------------GROUP------------------------------------------------")
```
print?

rmehta commented on frappe/www/unsubscribed.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def get_context(context):
+		if "user_email" in frappe.form_dict:
```
check indenting

rmehta reviewed (changes requested):
please share sceenshots

rmehta commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,81 @@
+{% extends "templates/web.html" %}
+
+{% block title %} {{_("Reset Password")}} {% endblock %}
```
title

rmehta commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,81 @@
+
+{% block page_content %}
+
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
```
pull from local!

Anurag810 replied:
done

rmehta commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,81 @@
+<script>
+$(document).ready(function(){
+	$("#select").click(function(){
+		if($("#select").prop("checked") == true){
```
not required!

rmehta commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,81 @@
+	</div>
+<hr>
+	<form method="POST" action="unsubscribed">
+			<input type="hidden" name="user_email" value="{{email}}">
```
indenting

Anurag810 replied:
resolved indentation problem

Commit:
bux fixes

Commit:
confirmed_unsubscribe code raw mysql query removed

Anurag810 commented:
Before 

![before](https://user-images.githubusercontent.com/32095923/48533914-4dbcee80-e8cc-11e8-8953-43a46b7cce89.png)

After

![after](https://user-images.githubusercontent.com/32095923/48533973-81981400-e8cc-11e8-949b-eec872149c59.gif)

For invalid or unsigned Request
![screenshot 2018-11-15 at 11 50 08 am](https://user-images.githubusercontent.com/32095923/48534020-b2784900-e8cc-11e8-95fd-984784c448fb.png)
or invalid requests

Commit:
moved everything to single file for optimisation

Commit:
UI fixes

Commit:
fixed csrf token not found

Commit:
removed the older function(method)

Commit:
code refracted or added some comment

Commit:
changes

Commit:
ui enhancement

Commit:
codacy fixes

Anurag810 commented:
before

![before](https://user-images.githubusercontent.com/32095923/48600007-2b8ca480-e990-11e8-91af-83e05c2d7ab6.png)

After

![after-new](https://user-images.githubusercontent.com/32095923/48600000-2891b400-e990-11e8-9ff8-7ae4cbe80fec.gif)

For unsigned and invalid request

![screenshot 2018-11-16 at 11 03 42 am](https://user-images.githubusercontent.com/32095923/48600012-2f202b80-e990-11e8-8914-1353368217dc.png)

Anurag810 commented:
final ui 

![screenshot 2018-11-16 at 11 03 42 am](https://user-images.githubusercontent.com/32095923/48606461-aa400c80-e9a5-11e8-9e9a-186429fe422a.png)

Commit:
Ui fixes

Anurag810 commented on frappe/email/doctype/newsletter/newsletter.py:
```diff
@@ -132,33 +132,15 @@ def get_email_groups(name):
+		data = frappe.get_doc('Email Group Member', doc[0].name)
+		data.unsubscribed = 1
+		data.save(ignore_permissions=True)
+		frappe.db.commit()
```
Done

Anurag810 replied:
removed frappe.db.commit

Anurag810 commented on frappe/www/unsubscribe.py:
```diff
@@ -0,0 +1,39 @@
+
+def get_email_groups(user_email):
+	# Return the all email_groups in which the email has been registered.
+	return frappe.get_all("Email Group Member", fields = ["email_group"], filters = {"email": user_email, "unsubscribed": 0}, ignore_permissions=True)
```
removed ignore permissions

Anurag810 commented on frappe/www/unsubscribe.py:
```diff
@@ -0,0 +1,39 @@
+
+def get_current_groups(name):
+	# Return current group by which the mail has been sent.
+	return frappe.db.get_all("Newsletter Email Group", ["email_group"],{"parent":name, "parenttype":"Newsletter"}, ignore_permissions=True)
```
removed ignore permission

Anurag810 commented on frappe/email/doctype/newsletter/newsletter.py:
```diff
@@ -132,33 +132,15 @@ def get_email_groups(name):
+def confirmed_unsubscribe(email, group):
+	""" unsubscribe the email(user) from the mailing list(email_group) """
+	frappe.flags.ignore_permissions=True
+	doc = frappe.get_all('Email Group Member', filters={"email": email, "email_group": group}, fields=['name','unsubscribed'], ignore_permissions = True)
```
ok, removed the ignore_permission

Anurag810 replied:
Actually, I think `get_doc` need two parameters `doctype` and `name ` to load specific doc (As given in documentation) .
So ,used `get_all` to fetch the name and used it to fetch and update doctype docs.

Alchez replied:
`get_doc` also takes `doctype` and `filters` as its arguments.

For example:

```python
user = frappe.get_doc("User", {"first_name": "Test"})
```

will return the User object if it exists in the system, otherwise throw a `DoesNotExistError`.

Anurag810 replied:
thanks, done

Commit:
Applied Requested changes

Commit:
code optimisation

surajshetty3416 commented:
@Anurag810 tests failing.

Also, is this still WIP?

Anurag810 commented:
It is completed please take a review once again and suggest changes to made.

Anurag810 commented:
looking where it is failing

Commit:
fixed unit test

Commit:
more validation

Commit:
Ui fixes

Commit:
removed print statement which was used for tracing

Commit:
change

surajshetty3416 commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,120 @@
+									<label><input type="checkbox" checked class="group" name='{{group.email_group}}'><span style="padding-left: 10px">{{group.email_group}}</span></label>
+								</div>
+							{% else %}
+								<div class="checkbox {% if email_groups|length > 20 %} col-sm-6 {% endif %}">
```
The code looks redundant here.

Anurag810 replied:
fixed

surajshetty3416 reviewed (changes requested):
Please update screenshots.

Anurag810 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into email-fixes

Anurag810 commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,120 @@
+			<div class='page-card-head'>
+				<span class='indicator blue'>Unsubscribe</span>
+			</div>
+			{% if email_groups != "none" %}
```
i have checked it is not working in jinja

Anurag810 commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,120 @@
+		.checkbox-container {
+			margin-top: 20px;
+		}
+		.main-div{
```
done

Anurag810 commented on frappe/www/unsubscribe.html:
```diff
@@ -0,0 +1,120 @@
+	{% endif %}
+
+	<style>
+	.hero-and-content {
```
oh i forget to remove it

Anurag810 committed:
done changes requested

Anurag810 committed:
done chnages requested

Anurag810 commented:
Updated screenshots

confirmation-page

![confirmation-page](https://user-images.githubusercontent.com/32095923/49069447-46dc9700-f24f-11e8-8dde-f8b2bc0c9f2e.png)

Confirmed Unsubscription page

![confirmed-unsubcription](https://user-images.githubusercontent.com/32095923/49069501-67a4ec80-f24f-11e8-833a-836c67cfe3c7.png)

if no subscription

![no-subs](https://user-images.githubusercontent.com/32095923/49069520-768b9f00-f24f-11e8-985a-ecf6eeb21919.png)


unsigned or invalid request


![invalid unsigned-request](https://user-images.githubusercontent.com/32095923/49069543-85725180-f24f-11e8-85bf-319f3b149d31.png)

Anurag810 commented:
Updated screenshots and done changes requested check it once again

surajshetty3416 committed:
Make code more self-explanatory

surajshetty3416 committed:
Fix some code formatting
- Add more descriptive name/id for buttons
- Rearrange a comment
- Remove some unwanted line breaks
- change status values

surajshetty3416 committed:
Merge branch 'develop' into Anurag810-email-fixes
